### PR TITLE
support for allowing custom variable based request headers

### PR DESCRIPTION
**Pull request by khgandhi**
This PR is a new feature that allows the user to configure the custom request headers names for certain fields like CLIENT_IP. The feature leverages the "request_headers_to_add" property.

GH Ticket https://github.com/lyft/envoy/issues/1016

### CODE REVIEW

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
   return hash;
 }
 
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
+	std::string variable_name;
```
Comment (htuch) [inline source/common/router/config_impl.cc]: Please run this through `fix_format`, this definitely doesn't comply with the style guide (2 spaces, etc.). CI won't allow this to pass until it's compliant.

File: include/envoy/http/access_log.h [hunk 2]
```diff
@@ -164,7 +175,7 @@ class Formatter {
                              const RequestInfo& request_info) const PURE;
 };
 
-typedef std::unique_ptr<Formatter> FormatterPtr;
+typedef std::shared_ptr<Formatter> FormatterPtr;
```
Comment (htuch) [inline include/envoy/http/access_log.h]: Also, is there a good reason this is now shared? Shared ownership semantics are more expensive and harder to reason about, so all things being equal it's best to keep unique. If there is a reason, please share it in some comment.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
...
+	if (format.find("%") == 0) {
```
Comment (htuch) [inline source/common/router/config_impl.cc]: It feels like we're duplicating a bunch of parsing code from AccessLogFormatParser::parse here for the icky parsing of the % delimited variables. Can we just do this in one place and reuse?

File: include/envoy/http/access_log.h [hunk 4]
```diff
@@ -114,6 +114,17 @@ class RequestInfo {
    * Set whether the request is a health check request or not.
    */
   virtual void healthCheck(bool is_hc) PURE;
+
+  /**
+   * Set the downstream address
+   */
+  virtual void downStreamAddress(std::string address) PURE;
```
Comment (mattklein123) [inline include/envoy/http/access_log.h]: Setting aside the formatting of getter/setter, I don't think there should be a setter on this interface for this. It doesn't really make sense as the address is fixed. Unclear why it would need to be changed.

File: source/common/router/config_impl.cc [hunk 5]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
   return hash;
 }
 
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: Can we please move all of this code into a new .cc/.h file. This is pretty self contained and this file is getting too huge as it is.

File: include/envoy/http/access_log.h (hunk 4 above)
Comment (khgandhi) [inline include/envoy/http/access_log.h]: @htuch i can definitely change to `setDownstreamAddress`.

File: include/envoy/http/access_log.h [hunk 6]
```diff
@@ -114,6 +114,17 @@ class RequestInfo {
...
+
+  /**
+   * Get the down stream address.
+   */
+  virtual std::string downStreamAddress() const PURE;
```
Comment (khgandhi) [inline include/envoy/http/access_log.h]: i will change that

File: include/envoy/http/access_log.h (hunk 2 above)
Comment (khgandhi) [inline include/envoy/http/access_log.h]: @htuch this not required anymore.. i will roll it back

File: source/common/router/config_impl.cc (hunk 5 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: @mattklein123 ok.. i will work on that refactoring.

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: @htuch the AccessLogFormatParser::parse supports a lot of formatting like parsing the request and response headers, supports multiple variables in a single format string.
For the custom request header, we do not have request header because the upstream cluster already gets all of that.
So it is a very small piece of parsing code that seems to be common between these parsers and hence kept it separate.

File: source/common/router/config_impl.cc (hunk 1 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: @htuch will definitely format it out to spec.

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (htuch) [inline source/common/router/config_impl.cc]: Why not just have the general AccessLogFormatParser::parse and then make this a special case? In general, I'm a fan of less code and less boiler plate. If we find a bug in parsing, I rather fix it in one place than two.

File: source/common/http/access_log/request_info_impl.h [hunk 7]
```diff
@@ -46,6 +46,13 @@ struct RequestInfoImpl : public RequestInfo {
 
   void healthCheck(bool is_hc) override { hc_request_ = is_hc; }
 
+  void downStreamAddress(std::string address) override {downstream_address_ = address;}
+
+    /**
+     * Get the down stream address.
+     */
+   std::string downStreamAddress() const override {return downstream_address_;};
```
Comment (RomanDzhabarov) [inline source/common/http/access_log/request_info_impl.h]: const reference here

File: source/common/http/conn_manager_impl.cc [hunk 8]
```diff
@@ -483,6 +483,9 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
 
   // Set the trusted address for the connection by taking the last address in XFF.
   downstream_address_ = Utility::getLastAddressFromXFF(*request_headers_);
+  log().debug("setting down stream address '{}' on request info",
```
Comment (RomanDzhabarov) [inline source/common/http/conn_manager_impl.cc]: remove this after testing

File: source/common/router/config_impl.cc [hunk 9]
```diff
@@ -7,6 +7,7 @@
 #include <regex>
 #include <string>
 #include <vector>
+#include <iostream>
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: alpha order

File: source/common/router/config_impl.cc [hunk 10]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
...
+	}else {
+		HeaderFormatterSharedPtr plain_header_formatter_shard_ptr (new PlainHeaderFormatter(format));
+		return plain_header_formatter_shard_ptr;
+	}
+	HeaderFormatterSharedPtr request_header_formatter_shard_ptr(new RequestHeaderFormatter(variable_name));
+	return request_header_formatter_shard_ptr;
+}
+
+RequestHeaderParserSharedPtr RequestHeaderParser::parse(const Json::Object& config) {
+	RequestHeaderParser* parser = new RequestHeaderParser();
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: do not use raw ptr

File: source/common/router/config_impl.cc [hunk 11]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
...
+	std::string variable_name;
+	if (format.find("%") == 0) {
+		size_t last_occ_pos = format.rfind("%");
+		if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+			throw EnvoyException(fmt::format("Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",format));
+		}
+		variable_name = format.substr(1, last_occ_pos - 1);
+
+
+	}else {
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: can you run fix_format

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: +1 to reuse the logic

File: source/common/router/config_impl.h [hunk 12]
```diff
@@ -173,6 +182,89 @@ class HashPolicyImpl : public HashPolicy {
...
+ * a formatter that expands the request header variable to a value based on info in RequestInfo.
+ */
+class RequestHeaderFormatter: public HeaderFormatter,
+							  Logger::Loggable<Logger::Id::config>  {
+
+public:
+	virtual ~RequestHeaderFormatter(){};
+	RequestHeaderFormatter(std::string& field_name);
+
+	std::string format(const Envoy::Http::AccessLog::RequestInfo& request_info) const;
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.h]: specify where override comes from and also add override at the end.

File: source/common/router/config_impl.h [hunk 13]
```diff
@@ -173,6 +182,89 @@ class HashPolicyImpl : public HashPolicy {
...
+	 virtual ~RequestHeaderParser(){};
+
+	 static RequestHeaderParserSharedPtr parse(const Json::Object& config);
+
+	 static HeaderFormatterSharedPtr parseInternal(const std::string& format);
+
+	 void evaluateRequestHeaders(Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+	   							  const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+	 std::map<Http::LowerCaseString, HeaderFormatterSharedPtr> headerFormatterMap() {return header_formatter_map_;};
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.h]: any reason to have ordered headers here?

File: source/common/router/config_impl.cc (hunk 11 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: ok..

File: source/common/router/config_impl.cc (hunk 9 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: ok

File: source/common/http/conn_manager_impl.cc (hunk 8 above)
Comment (khgandhi) [inline source/common/http/conn_manager_impl.cc]: ok

File: source/common/http/access_log/request_info_impl.h (hunk 7 above)
Comment (khgandhi) [inline source/common/http/access_log/request_info_impl.h]: const reference is because this is a getter and is not intended to change any member variable value on the class..

File: source/common/router/config_impl.h (hunk 12 above)
Comment (khgandhi) [inline source/common/router/config_impl.h]: ok..

File: source/common/router/config_impl.cc (hunk 10 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: ok.. i will address that.

File: source/common/http/access_log/request_info_impl.h (hunk 7 above)
Comment (RomanDzhabarov) [inline source/common/http/access_log/request_info_impl.h]: i meant to say, const std::string& downstreamAddress() ....

File: source/common/router/config_impl.cc [hunk 14]
```diff
@@ -65,6 +67,82 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
...
+	if (field_name == "PROTOCOL") {
+	    field_extractor_ = [](const Envoy::Http::AccessLog::RequestInfo& request_info) {
+	      return Envoy::Http::AccessLog::AccessLogFormatUtils::protocolToString(request_info.protocol());
+	    };
+	 }else if (field_name == "CLIENT_IP") {
+		 field_extractor_ = [](const Envoy::Http::AccessLog::RequestInfo& request_info) {
+		  log().debug("field extractor invoked");
+		  return request_info.downStreamAddress();
+		};
+	 }else {
```
Comment (jwfang) [inline source/common/router/config_impl.cc]: it would be nice to have a CLIENT_PORT here

File: source/common/router/config_impl.cc (hunk 10 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: @RomanDzhabarov here the raw pointer is converted to a share ptr below at line 100[here](https://github.com/khgandhi/envoy/blob/1cac1769a4119fe1b7eea813c503311f2c750f4d/source/common/router/config_impl.cc#L100).

Comment (dnoe) [inline source/common/router/config_impl.cc]: In this use case you probably want to use std::make_shared<RequestHeaderParser>.  This is the safest way to use shared_ptr, and is exception safe (if an exception is thrown in the code between line 89 and line 100, there will be a leak).

**Comment by RomanDzhabarov** [comment]
@khgandhi can you make tests passing?

**Comment by khgandhi** [comment]
@RomanDzhabarov i have fixed the tests. can you please take another pass at it ?

**Comment by RomanDzhabarov** [comment]
thanks I'll look today

File: bazel/README.md [hunk 15]
```diff
@@ -114,6 +114,44 @@ you can run the test with `--strategy=TestRunner=standalone`, e.g.:
 ```
 bazel test //test/common/http:async_client_impl_test --strategy=TestRunner=standalone --run_under=/some/path/foobar.sh
 ```
+
+# Code Formatter with Bazel
+Envoy code is formatted based on [`clang-format-3.6`](http://releases.llvm.org/3.6.0/tools/clang/docs/ClangFormatStyleOptions.html).
```
Comment (htuch) [inline bazel/README.md]: Thanks for adding this writeup. Can you move it to a separate PR?

File: source/common/router/config_impl.h [hunk 16]
```diff
@@ -39,6 +42,7 @@ class Matchable {
                                       uint64_t random_value) const PURE;
 };
 
+// forward declaration
```
Comment (htuch) [inline source/common/router/config_impl.h]: This comment is probably not needed as the code speaks for itself.

File: source/common/router/req_header_formatter.cc [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  return request_header_parser;
+}
+
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
+
+  for (const std::pair<Http::LowerCaseString, std::string>& to_add : requestHeadersToAdd) {
+
+    log().debug("request headers key {}", to_add.first.get());
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: Please switch to the new `ENVOY_LOG` style of logging here and elsewhere.

File: source/common/router/req_header_formatter.cc [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  return request_header_parser;
+}
+
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
+
+  for (const std::pair<Http::LowerCaseString, std::string>& to_add : requestHeadersToAdd) {
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: Nit: this could be written more concisely as `const auto& to_add : requestHeadersToAdd`.

File: source/common/router/req_header_formatter.cc [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    variable_name = format.substr(1, last_occ_pos - 1);
+
+  } else {
+    HeaderFormatterSharedPtr plain_header_formatter_shard_ptr(new PlainHeaderFormatter(format));
+    return plain_header_formatter_shard_ptr;
+  }
+  HeaderFormatterSharedPtr request_header_formatter_shard_ptr(
+      new RequestHeaderFormatter(variable_name));
+  return request_header_formatter_shard_ptr;
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: Nit: I think you can drop the `_shard_ptr` suffix (it's shared right, not shard?).

File: source/common/router/req_header_formatter.cc [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
+  std::string variable_name;
+  if (format.find("%") == 0) {
+    size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format(
+          "Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",
+          format));
+    }
+    variable_name = format.substr(1, last_occ_pos - 1);
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: Why do we only support a single variable? I had in mind that this could be equally powerful to the access log formatter (and reuse its parsing code).

File: source/common/router/req_header_formatter.cc [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+    log().debug("request headers key {}", to_add.first.get());
+
+    // search if the request has formatters defined
+    if (header_formatter_map_.count(to_add.first) == 1) {
+
+      auto search = header_formatter_map_.find(to_add.first);
+      // access the formatter
+      if (search != header_formatter_map_.end()) {
+        std::string formatted_header_value = search->second->format(requestInfo);
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: Nit: `const std::string`

File: source/common/router/req_header_formatter.cc [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+  for (const std::pair<Http::LowerCaseString, std::string>& to_add : requestHeadersToAdd) {
+
+    log().debug("request headers key {}", to_add.first.get());
+
+    // search if the request has formatters defined
+    if (header_formatter_map_.count(to_add.first) == 1) {
+
+      auto search = header_formatter_map_.find(to_add.first);
+      // access the formatter
+      if (search != header_formatter_map_.end()) {
```
Comment (htuch) [inline source/common/router/req_header_formatter.cc]: There shouldn't be an `if` here, you tested the condition in the outer `if`. I would rewrite it to do the `find`, and then just condition on if `!= ..end()` on the outer `if`.

File: source/common/router/req_header_formatter.h [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+
+  static RequestHeaderParserSharedPtr parse(const Json::Object& config);
+
+  static HeaderFormatterSharedPtr parseInternal(const std::string& format);
+
+  void evaluateRequestHeaders(
+      Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+  std::map<Http::LowerCaseString, HeaderFormatterSharedPtr> headerFormatterMap() {
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.h]: why not unordered_map?

File: test/common/http/access_log/access_log_impl_test.cc [hunk 24]
```diff
@@ -65,6 +65,10 @@ class TestRequestInfo : public RequestInfo {
   bool healthCheck() const override { return hc_request_; }
   void healthCheck(bool is_hc) override { hc_request_ = is_hc; }
 
+  std::string downstream_address_;
```
Comment (RomanDzhabarov) [inline test/common/http/access_log/access_log_impl_test.cc]: move this down below

File: test/common/http/access_log/access_log_impl_test.cc [hunk 25]
```diff
@@ -65,6 +65,10 @@ class TestRequestInfo : public RequestInfo {
...
+  void setDownstreamAddress(std::string address) override { downstream_address_ = address; }
```
Comment (RomanDzhabarov) [inline test/common/http/access_log/access_log_impl_test.cc]: can you initialize RequestInfo with DownstreamAddress in the ctor?

File: test/common/router/config_impl_test.cc [hunk 26]
```diff
@@ -1996,5 +2001,219 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
   EXPECT_TRUE(config_ptr->route(headers, 0)->routeEntry()->includeVirtualHostRateLimits());
 }
 
+TEST(CustomRequestHeadersTest, AddNewHeader) {
+  std::string json = R"EOF(
```
Comment (RomanDzhabarov) [inline test/common/router/config_impl_test.cc]: const

File: test/common/router/config_impl_test.cc [hunk 27]
```diff
@@ -1996,5 +2001,219 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
...
+      "value": "%CLIENT_IP"
+    }
+  ]
+}
+	  )EOF";
+  Json::ObjectSharedPtr loader = Json::Factory::loadFromString(json);
+  NiceMock<Runtime::MockLoader> runtime;
+  NiceMock<Upstream::MockClusterManager> cm;
+  NiceMock<Envoy::Http::AccessLog::MockRequestInfo> requestInfo;
+  EXPECT_THROW(ConfigImpl config(*loader, runtime, cm, true), EnvoyException);
```
Comment (RomanDzhabarov) [inline test/common/router/config_impl_test.cc]: can you validate the message? so we are sure that EnvoyException is raised because of the wrong format in the headers

File: source/common/router/config_impl.cc [hunk 28]
```diff
@@ -176,19 +180,17 @@ bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t ran
 
 const std::string& RouteEntryImplBase::clusterName() const { return cluster_name_; }
 
-void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers,
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: does it have to be a reference to RequestInfo? instead of const ref

File: source/common/http/access_log/request_info_impl.h [hunk 29]
```diff
@@ -46,6 +46,13 @@ struct RequestInfoImpl : public RequestInfo {
 
   void healthCheck(bool is_hc) override { hc_request_ = is_hc; }
 
+  void setDownstreamAddress(std::string address) override { downstream_address_ = address; }
+
+  /**
+   * Get the down stream address.
```
Comment (RomanDzhabarov) [inline source/common/http/access_log/request_info_impl.h]: one word

File: source/common/router/config_impl.cc [hunk 30]
```diff
@@ -2,6 +2,7 @@
 
 #include <chrono>
 #include <cstdint>
+#include <iostream>
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: do not see why this is in here

File: source/common/router/config_impl.h [hunk 31]
```diff
@@ -2,6 +2,7 @@
 
 #include <chrono>
 #include <cstdint>
+#include <iostream>
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.h]: ditto

File: source/common/router/req_header_formatter.cc (hunk 20 above)
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: +1 to the @htuch

File: source/common/router/req_header_formatter.cc [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+RequestHeaderParserSharedPtr RequestHeaderParser::parse(const Json::Object& config) {
+
+  RequestHeaderParserSharedPtr request_header_parser(new RequestHeaderParser());
+  if (config.hasObject("request_headers_to_add")) {
+    std::vector<Json::ObjectSharedPtr> request_headers =
+        config.getObjectArray("request_headers_to_add");
+    for (const Json::ObjectSharedPtr& header : request_headers) {
+      if (!header->getString("key").empty() && !header->getString("value").empty()) {
+        log().debug("adding key {} to header formatter map", header->getString("key"));
+        request_header_parser->header_formatter_map_.emplace(
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: i think you could do, .emplace(Http::LowerCaseString(header->getString("key")), RequestHeaderParser::parseInternal(header->getString("value")))

File: source/common/router/req_header_formatter.cc [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+      }
+    }
+  }
+
+  return request_header_parser;
+}
+
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: why this is std::list and not a std::vector?

File: source/common/router/req_header_formatter.cc [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+        std::string formatted_header_value = search->second->format(requestInfo);
+        headers.addStaticKey(to_add.first, formatted_header_value);
+      }
+    } else {
+      headers.addStatic(to_add.first, to_add.second);
+    }
+  }
+}
+
+RequestHeaderFormatter::RequestHeaderFormatter(std::string& field_name) {
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: const std::string&

File: source/common/router/req_header_formatter.cc [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+      return Envoy::Http::AccessLog::AccessLogFormatUtils::protocolToString(
+          request_info.protocol());
+    };
+  } else if (field_name == "CLIENT_IP") {
+    field_extractor_ = [](const Envoy::Http::AccessLog::RequestInfo& request_info) {
+      return request_info.getDownstreamAddress();
+    };
+  } else {
+    throw EnvoyException(
+        fmt::format("field {} not supported as custom request header", field_name));
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: field '{}' is not supported ....

File: source/common/router/req_header_formatter.h [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+  virtual ~RequestHeaderFormatter(){};
+  RequestHeaderFormatter(std::string& field_name);
+
+  // HeaderFormatter::format
+  std::string format(const Envoy::Http::AccessLog::RequestInfo& request_info) const override;
+
+private:
+  std::function<std::string(const Envoy::Http::AccessLog::RequestInfo&)> field_extractor_;
+};
+
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.h]: kill line

File: source/common/router/req_header_formatter.cc (hunk 20 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: as per @mattklein123 comment on the ticket, req parameters that are sent by client's already get added to the upstream request as headers (These are one that access log formatter gets it via the `REQ(X?Y):Z%` prefix. A lot of the response headers might not be applicable here (bytes received, duration, etc.) as well. Hence, we might need to provide a set of parameters that can be used as headers to the upstream requests. to begin with, i have started with CLIENT_IP and PROTOCOL that might be something that upstream clusters need it.

File: source/common/router/req_header_formatter.cc (hunk 19 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: yes.. i will address it.

File: source/common/router/config_impl.h (hunk 16 above)
Comment (khgandhi) [inline source/common/router/config_impl.h]: yes.. will remove it

File: bazel/README.md (hunk 15 above)
Comment (khgandhi) [inline bazel/README.md]: sure. i will have a separate PR for this.

File: source/common/router/req_header_formatter.cc (hunk 18 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: ok.. i will address that

File: source/common/router/config_impl.cc (hunk 30 above)
Comment (khgandhi) [inline source/common/router/config_impl.cc]: haha.. i added it for debugging for std::cout.. i will remove it

File: source/common/router/req_header_formatter.cc (hunk 33 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: requestHeadersToAdd is pulled from configuration and is already present as a list in the code.

**Comment by RomanDzhabarov** [comment]
I'll review this once comments addressed and master merged.

**Comment by khgandhi** [comment]
@RomanDzhabarov @htuch i have uploaded a new commit and incorporated all the feedback.. Please review.

File: include/envoy/http/header_map.h [hunk 37]
```diff
@@ -30,13 +30,20 @@ class LowerCaseString {
 
   const std::string& get() const { return string_; }
   bool operator==(const LowerCaseString& rhs) const { return string_ == rhs.string_; }
+  bool operator<(const LowerCaseString& rhs) const { return string_ < rhs.string_; }
```
Comment (RomanDzhabarov) [inline include/envoy/http/header_map.h]: there is no need to this anymore as unordered_map is not using '<' operator

File: include/envoy/http/access_log.h [hunk 38]
```diff
@@ -114,6 +114,16 @@ class RequestInfo {
    * Set whether the request is a health check request or not.
    */
   virtual void healthCheck(bool is_hc) PURE;
+
+  /**
+   * Set the downstream address
+   */
+  virtual void setDownstreamAddress(std::string address) PURE;
```
Comment (RomanDzhabarov) [inline include/envoy/http/access_log.h]: pass `address` argument by const reference

Comment (mattklein123) [inline include/envoy/http/access_log.h]: I am still not in favor of this being part of the interface at all. Why is this required? (This is just a quick comment I have not done a full review).

File: test/common/router/config_impl_test.cc [hunk 39]
```diff
@@ -2029,5 +2034,229 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
   EXPECT_TRUE(config_ptr->route(headers, 0)->routeEntry()->includeVirtualHostRateLimits());
 }
 
+TEST(CustomRequestHeadersTest, AddNewHeader) {
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: Another quick comment. Please move the majority of the functional tests into a new test file to match the formatter impl. You can have a quick test here if you want just to verify that things are called correctly.

File: include/envoy/http/access_log.h (hunk 38 above)
Comment (khgandhi) [inline include/envoy/http/access_log.h]: @mattklein123 we are now passing the requestinfo to the finalizeRequestHeaders method so that it can access to information like protocol and other data about the request. (similar to access log formatter). Down stream address is added to the request info so that request_header_formatter.cc can insert CLIENT_IP into the request to the upstream clusters.

Comment (mattklein123) [inline include/envoy/http/access_log.h]: But there is no reason to provide a setter on the interface. When will it ever be set at an interface level?

Comment (khgandhi) [inline include/envoy/http/access_log.h]: yes.. it is set in router.cc before you call the finalizeRequestHeaders method to set the downstream address.

Comment (mattklein123) [inline include/envoy/http/access_log.h]: This should be set by the connection manager since it knows the downstream address. There is no reason for the router to set it at all.

File: test/common/router/config_impl_test.cc (hunk 39 above)
Comment (khgandhi) [inline test/common/router/config_impl_test.cc]: the two test cases added are inline with other tests in config_impl_test.cc where we have a simple virtual hosts config and we call the route logic with mock headers.  Do you just want me to create a separate test file like req_header_formatter_test.cc and move these two test cases in there ?

Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: It would be good to have specific tests for the req_header_formatter code that have nothing to do with the route config, since the formatter will get more complicated in the future. so you should have two layers of tests.

File: include/envoy/http/access_log.h (hunk 38 above)
Comment (khgandhi) [inline include/envoy/http/access_log.h]: You mean to say add something like follows 
request_info_.setDownstreamAddress(downstream_address_) in conn_manager_impl.cc.decodeHeaders [here](https://github.com/lyft/envoy/blob/master/source/common/http/conn_manager_impl.cc#L484)

File: test/common/router/config_impl_test.cc (hunk 39 above)
Comment (khgandhi) [inline test/common/router/config_impl_test.cc]: ok understood.. i will keep the existing test cases as is and add more for the req_header_formatter in a separate test file.

File: source/common/http/access_log/request_info_impl.h (hunk 29 above)
Comment (RomanDzhabarov) [inline source/common/http/access_log/request_info_impl.h]: this is still not addressed ;)

File: source/common/router/req_header_formatter.cc [hunk 40]
```diff
@@ -0,0 +1,92 @@
...
+    return plain_header_formatter_shared_ptr;
+  }
+  HeaderFormatterSharedPtr request_header_formatter_shared_ptr(
+      new RequestHeaderFormatter(variable_name));
+  return request_header_formatter_shared_ptr;
+}
+
+RequestHeaderParserSharedPtr RequestHeaderParser::parse(const Json::Object& config) {
+
+  RequestHeaderParserSharedPtr request_header_parser(new RequestHeaderParser());
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: why does it have to be shared ptr and not just RequestHeaderParser?

File: source/common/router/req_header_formatter.cc [hunk 41]
```diff
@@ -0,0 +1,92 @@
...
+    if (search != header_formatter_map_.end()) {
+      const std::string formatted_header_value = search->second->format(requestInfo);
+      headers.addStaticKey(to_add.first, formatted_header_value);
+    } else {
+      headers.addStatic(to_add.first, to_add.second);
+    }
+  }
+}
+
+RequestHeaderFormatter::RequestHeaderFormatter(std::string& field_name) {
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.cc]: Instead of this why don't reuse https://github.com/lyft/envoy/blob/master/source/common/http/access_log/access_log_formatter.cc#L229 ?

File: source/common/router/req_header_formatter.h [hunk 42]
```diff
@@ -0,0 +1,100 @@
...
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterSharedPtr;
+
+/**
+ * a formatter that expands the request header variable to a value based on info in RequestInfo.
+ */
+class RequestHeaderFormatter : public HeaderFormatter, Logger::Loggable<Logger::Id::config> {
+
+public:
+  virtual ~RequestHeaderFormatter(){};
+  RequestHeaderFormatter(std::string& field_name);
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.h]: const in the majority of the places where you supply arg by reference and do not have an intent to modify that.

File: source/common/router/req_header_formatter.h [hunk 43]
```diff
@@ -0,0 +1,100 @@
...
+    static_value_ = static_header_value;
+  };
+
+  // HeaderFormatter::format
+  std::string format(const Envoy::Http::AccessLog::RequestInfo&) const override {
+    return static_value_;
+  };
+
+private:
+  std::string static_value_;
```
Comment (RomanDzhabarov) [inline source/common/router/req_header_formatter.h]: const

File: source/common/router/req_header_formatter.cc (hunk 41 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: none of the fields that the access_log_formatter currently support are applicable for request_header_formatter use case. Also, the access log formatter supports parsing any request or response header by using "REQ(.." or "RESP(.." prefix in the formatter with is also required in request_header_formatter as well.

That is why we have a separate formatter for request_header formatter.

File: source/common/router/req_header_formatter.h (hunk 43 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.h]: i will address the const ref on setDownstream address and other places in my next feedback set :)

**Comment by mattklein123** [comment]
@khgandhi friendly ping, can we merge master and get this passing tests so we can move this along?

**Comment by mattklein123** [comment]
I'm going to go ahead and close this for now. Feel free to reopen when you are ready to keep going on it.

**Comment by khgandhi** [comment]
@mattklein123 would like to reopen this issue since i rebased the PR with the master

**Comment by mattklein123** [comment]
@khgandhi for some reason GH is not letting me reopen this PR. I don't know why. Can you make a new one?

**Comment by khgandhi** [comment]
@mattklein123 ok.. i will squash all the commits into one and create a new PR

File: source/common/router/req_header_formatter.cc (hunk 40 above)
Comment (khgandhi) [inline source/common/router/req_header_formatter.cc]: @RomanDzhabarov fixed this in a different PR #1390
